Upload large file to blob storage block by block

GeraldSchmitt opened the pull request:
# Checklist
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-c/blob/master/.github/CONTRIBUTING.md).
- [x] _The tests are partially written, and I'm working on it. So far, the already existing tests impacted by the new code have been updated, and some tests have been written to cover the 3 main functions `IoTHubClient_LL_LargeFileOpen_Impl`, `IoTHubClient_LL_LargeFileWrite_Impl`, `IoTHubClient_LL_LargeFileClose_Impl`. The coverage should be close to 100%._
- If this is a modification that impacts the behavior of a public API: _Preexisting public APIs have not been modified, but a new API has been added._
  - [x] I edited the corresponding document in the `devdoc` folder and added or modified requirements. _Not yet_
- I submitted this PR against the correct branch: 
  - [X] This pull-request is submitted against the `master` branch. _I think so._

# Reference/Link to the issue solved with this PR (if any)
[Issue 211: Upload large file to IoTHub blob storage with azure-iot-sdk-c](https://github.com/Azure/azure-iot-sdk-c/issues/211)

# Description of the problem
We need to upload a big file to a blob storage. The problem is that the file does not fit into memory, so we need an API able to upload the file block by block.

# Description of the solution
The desired mecanism already exists but is hidden inside the function `IoTHubClient_LL_UploadToBlob`.
We can easily expose this mecanism with a new API, without impacting the existing function.

#### Existing calling sequence in pseudo code:
```C
IoTHubClient_LL_UploadToBlob_Impl(blobHandle, filename, source, size)
{
    // step A: prepare sasuri, ressources, certificates...

    Blob_UploadFromSasUri(...)
    {
        // step B: open connection, create blob...
        foreach n block of data generated from source
        {
            Blob_UploadNextBlock(block of data,...)
            {
                // step Cn: upload Block to server
            }
        }
        // step D: close blob, connection
    }

    // step E: dipose ressources...
}
```
Everything inside `IoTHubClient_LL_UploadToBlob_Impl` is hidden to the user, in particular the upload of a single block.

#### Added calling sequence in pseudo code, where the upload of a block is exposed to the user:
```C
/* OPEN CONNECTION */
largeFileHandle = IoTHubClient_LL_LargeFileOpen_Impl(blobHandle, filename)
{
    // Init largeFileHandle
    
    LARGE_FILE_upload_blob_start(...)
    {
        // step A: prepare sasuri, ressources, certificates...

        Blob_UploadFromSasUri_start(...)
        {
            // step B: open connection, create blob...
        }
    }
}

foreach n block of data generated by the user
{
    /* WRITE DATA */
    result = IoTHubClient_LL_LargeFileWrite_Impl(largeFileHandle, blockSource, blockSize)
    {
        Blob_UploadNextBlock(...)
        {
            // step Cn: upload Block to server
        }
    }
}

/* CLOSE CONNECTION */
result = IoTHubClient_LL_LargeFileClose_Impl(largeFileHandle)
{
    LARGE_FILE_upload_blob_stop(...)
    {
        Blob_UploadFromSasUri_stop(...)
        {
            // step D: close blob, connection
        }

        // step E: dispose ressources...       
    }

    // dispose largeFileHandle
}

```

#### New available API:
```C
IOTHUB_CLIENT_LARGE_FILE_HANDLE IoTHubClient_LargeFileOpen(IOTHUB_CLIENT_HANDLE handle, 
                                                           const char* destinationFileName);
														   
IOTHUB_CLIENT_RESULT IoTHubClient_LargeFileClose(IOTHUB_CLIENT_LARGE_FILE_HANDLE handle);

IOTHUB_CLIENT_RESULT IoTHubClient_LargeFileWrite(IOTHUB_CLIENT_LARGE_FILE_HANDLE fileHandle, 
                                                 const unsigned char* source,
                                                 size_t size);
```

# Possible improvements

 - The code can be refactored, so that `IoTHubClient_LL_UploadToBlob_Impl` uses this new API. This would reduce code dupplication.
 - The new function `IoTHubClient_LL_LargeFileWrite_Impl` fails if the data block is bigger than 4Mb. The function could be changed easily to upload the data in several blocks of 4Mb instead of failling.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/azure-iot-sdk-c?pullRequest=247) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/azure-iot-sdk-c?pullRequest=247) before we can accept your contribution.<br/><br/>:x: GeraldSchmitt [sign now](https://cla.opensource.microsoft.com/Azure/azure-iot-sdk-c?pullRequest=247)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/azure-iot-sdk-c?pullRequest=247) it.</sub>

tameraw commented:
@GeraldSchmitt - Thanks for your contribution.
We'll review the PR and provide you with feedback. Regards.

ewertons commented:
Hi @GeraldSchmitt ,

we are reviewing your changes. We will post comments and suggestions briefly.

Thanks for contributing to Azure IoT SDK!

GeraldSchmitt commented:
Hi @ewertons ,

This is great news. I'm waiting for your comments!

GeraldSchmitt commented:
Hi @ewertons ,
Thanks for the detailed review!
I will take the suggestions into account, and the API you suggest makes sense!

Some questions/remarks:
 - I think you made a copy/paste error in your answer though: the iothubclient module and iothubclient_ll module contains the same API description.
 - Do you use a tool to check unit test coverage, that I could use too?

ewertons commented:
Hi @GeraldSchmitt ,
thanks!

The APIs are correct. 
They look similar because iothubclient_ll and iothubclient are somehow similar, with the difference that the later is threaded, while the first is not (ll stands for low-layer or low-level).

Our design is as such:
iothubclient_ll: 
- contains all the functionalities of the Azure IoT device client SDK; 
- it has no threading support (to allow being run on low level devices); 
- requires customer application to constantly and frequently invoke IoTHubClient_LL_DoWork() for the client internal operations to be executed.

iothubclient:
- threaded version of the iothubclient_ll module;
- contains API functions that wrap up the iothubclient_ll module (adding up locks for thread safety);
- customer application **does not need** to invoke any _DoWork() for the client SDK to  function (all internal operations are triggered by a separate thread launched/controlled by the iothubclient module).

I would make one change on the design proposed though: please add the IOTHUB_CLIENT_FILE_UPLOAD_GET_DATA_CALLBACK callback definition only on iothubclient_ll.h, re-using it on iothubclient module.

ewertons commented:
We currently do not use a tool for code coverage, however we have plans to automate this.

Our code coverage is currently done through the "requirements" written in our requirement docs (e.g., [this](https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/devdoc/requirement_docs/iothubclient_requirements.md)), which must be present in both the code and the unit tests.

For memory leaks we use [valgrind](http://valgrind.org/), and for thread verification we use [helgrind](http://valgrind.org/info/tools.html#helgrind).

GeraldSchmitt commented:
Thanks for the details. It's clearer to me now!

ewertons commented:
You are absolutely correct. Kudos to you for how well you cracked the Azure IoT C SDK.

In this case we are keeping the similar definition of the functions in each module (iothubclient_ll and iothubclient) to make sure it is exposed properly on both.

GeraldSchmitt commented:
Hi @ewertons,

Thanks for your quick answer!

Just to be sure I understand everything:

The API you suggest, based on `IOTHUB_CLIENT_FILE_UPLOAD_GET_DATA_CALLBACK`, `IoTHubClient_UploadMultipleBlocksToBlobAsync` and `IoTHubClient_LL_UploadMultipleBlocksToBlob` is still valid, right?

This new API will use the functions `IoTHubClient_LL_LargeFileOpen_Impl`, `IoTHubClient_LL_LargeFileWrite_Impl` and `IoTHubClient_LL_LargeFileClose_Impl` I added in `iothub_client_ll_uploadtoblob.h`.

The API I defined  in `iothub_client.h` and `iothub_client_ll.h` will be removed and replaced with the API you suggest.

ewertons commented:
Very good, thanks for confirming.

The API you suggest, based on IOTHUB_CLIENT_FILE_UPLOAD_GET_DATA_CALLBACK, IoTHubClient_UploadMultipleBlocksToBlobAsync and IoTHubClient_LL_UploadMultipleBlocksToBlob is still valid, right?
> Yes, correct.

This new API will use the functions IoTHubClient_LL_LargeFileOpen_Impl, IoTHubClient_LL_LargeFileWrite_Impl and IoTHubClient_LL_LargeFileClose_Impl I added in iothub_client_ll_uploadtoblob.h.
> I believe we could remove the 3 APis and consolidate. I wouldn't be opposed to getting rid of the "_Impl" functions as well, adding all the functionality within the public APIs implementations (IoTHubClient_UploadMultipleBlocksToBlobAsync and IoTHubClient_LL_UploadMultipleBlocksToBlob, on their respective .c files).

GeraldSchmitt commented:
Yes, you're right, I think too there's a way to remove the 3 API.
I'll try to do that soon.
Thanks for your guidance!

GeraldSchmitt commented:
Here's some update:
The API using  `IOTHUB_CLIENT_FILE_UPLOAD_GET_DATA_CALLBACK` is now working. I removed all the previous "LARGE_FILE" API.
I did not get rid of the _impl functions yet, because it was easier for me stick to the existing architecture.
I replaced `Blob_UploadFromSasUri` by `Blob_UploadMultipleBlocksFromSasUri`.

The code is consolidated: `IoTHubClient_UploadToBlobAsync`, as well as `IoTHubClient_LL_UploadToBlob` both use `IoTHubClient_LL_UploadMultipleBlocksToBlob_Impl` internally.
The biggest change is that there is no more special case for blobs < 64MB: every blob is uploaded Block by Block. We didn't discuss this, but it seemed necessary to me in order to remove dupplication.

There is still some work to do:
 - The UT are broken (I'm fixing this)
 - Some error handling (especially block size and block count)
 - Remove `int* isError` parameter from `Blob_UploadNextBlock`
 - Update the specs
 - ...

Can you check that I understood your recommandation so far?

Also, one question: 
 - What if the user wants to upload a block bigger than 4MB when `IOTHUB_CLIENT_FILE_UPLOAD_GET_DATA_CALLBACK` is called? I think we should return an error, but we could also split the block internally. What do you think?

ewertons commented:
@GeraldSchmitt , 
I would like to let you know that I am tracking the commits you are making and I intend to review them soon. We are just fully booked resolving a couple critical issues, and as soon as we are done I will review your code. Thanks.

ewertons commented on iothub_client/src/iothub_client_ll_uploadtoblob.c:
```diff
@@ -761,9 +768,62 @@ static int IoTHubClient_LL_UploadToBlob_step3(IOTHUB_CLIENT_LL_UPLOADTOBLOB_HAND
+    return result;
+}
+
+IOTHUB_CLIENT_RESULT IoTHubClient_LL_UploadMultipleBlocksToBlob_Impl(IOTHUB_CLIENT_LL_UPLOADTOBLOB_HANDLE handle, const char* destinationFileName, IOTHUB_CLIENT_FILE_UPLOAD_GET_DATA_CALLBACK getDataCallback, void* context)
```
I believe some compilers might complain of the upper function calling another that hasn't been defined yet.
To be safe, I would recommend moving it above the IoTHubClient_LL_UploadToBlob_Impl

GeraldSchmitt replied:
I think this is true for functions defined in the same file. But here, `IoTHubClient_LL_UploadMultipleBlocksToBlob_Impl` is already defined in `iothub_client_ll_uploadtoblob.h`, so it should work with every compiler. Maybe some very exotic compilers could complain? Anyway, just to be safe, I moved it.

ewertons commented on iothub_client/src/iothub_client_ll_uploadtoblob.c:
```diff
@@ -774,10 +834,10 @@ IOTHUB_CLIENT_RESULT IoTHubClient_LL_UploadToBlob_Impl(IOTHUB_CLIENT_LL_UPLOADTO
         )
     {
-        LogError("invalid argument detected handle=%p destinationFileName=%p source=%p size=%zu", handle, destinationFileName, source, size);
+        LogError("invalid argument detected handle=%p destinationFileName=%p getDataCallback=%p context=%zu", handle, destinationFileName, getDataCallback, context);
```
Please do not print the context. Not needed.

GeraldSchmitt replied:
Done

ewertons commented on iothub_client/src/iothub_client_ll_uploadtoblob.c:
```diff
@@ -761,9 +768,62 @@ static int IoTHubClient_LL_UploadToBlob_step3(IOTHUB_CLIENT_LL_UPLOADTOBLOB_HAND
 {
     IOTHUB_CLIENT_RESULT result;
+
+    BLOB_UPLOAD_CONTEXT context;
```
You can define this variable within the else block. It doesn't need to show up until it's for sure needed.

GeraldSchmitt replied:
Done

ewertons commented on iothub_client/src/iothub_client_ll_uploadtoblob.c:
```diff
@@ -761,9 +768,62 @@ static int IoTHubClient_LL_UploadToBlob_step3(IOTHUB_CLIENT_LL_UPLOADTOBLOB_HAND
+
+    BLOB_UPLOAD_CONTEXT context;
+
+    if (source == NULL && size > 0)
```
Aren't there some checks missing for other invalid combinations?
Would this work? 
```c
if (source == NULL || size == 0)
```

GeraldSchmitt replied:
If this requirement is still valid 
`Codes_SRS_IOTHUBCLIENT_LL_02_063: [ If source is NULL and size is greater than 0 then IoTHubClient_LL_UploadToBlob shall fail and return IOTHUB_CLIENT_INVALID_ARG. ]`
then the condition is correct. It means it is allowed to create an empty file.

ewertons commented on iothub_client/src/iothub_client_ll_uploadtoblob.c:
```diff
@@ -761,9 +768,62 @@ static int IoTHubClient_LL_UploadToBlob_step3(IOTHUB_CLIENT_LL_UPLOADTOBLOB_HAND
+
+IOTHUB_CLIENT_RESULT IoTHubClient_LL_UploadMultipleBlocksToBlob_Impl(IOTHUB_CLIENT_LL_UPLOADTOBLOB_HANDLE handle, const char* destinationFileName, IOTHUB_CLIENT_FILE_UPLOAD_GET_DATA_CALLBACK getDataCallback, void* context)
+{
+    IOTHUB_CLIENT_RESULT result;
```
Besides `result`, please move all other variable definitions to within the block that uses it.
We don't define a variable before we are sure we will be able to use it (since you have a arg check that could fail).

GeraldSchmitt replied:
Done

GeraldSchmitt commented on iothub_client/inc/blob.h:
```diff
@@ -39,15 +46,28 @@ DEFINE_ENUM(BLOB_RESULT, BLOB_RESULT_VALUES)
 * @param	SASURI	        The URI to use to upload data
-* @param	size		    The size of the data to be uploaded (can be 0)
-* @param	source		    A pointer to the byte array to be uploaded (can be NULL, but then size needs to be zero)
+* @param	getDataCallback	A callback to be invoked to acquire the file chunks to be uploaded, as well as to indicate the status of the upload of the previous block.
```
Yep, that was definitely a bunch of rogue tabs!

GeraldSchmitt commented on iothub_client/inc/blob.h:
```diff
@@ -39,15 +46,28 @@ DEFINE_ENUM(BLOB_RESULT, BLOB_RESULT_VALUES)
-* @param	size		    The size of the data to be uploaded (can be 0)
-* @param	source		    A pointer to the byte array to be uploaded (can be NULL, but then size needs to be zero)
+* @param	getDataCallback	A callback to be invoked to acquire the file chunks to be uploaded, as well as to indicate the status of the upload of the previous block.
+* @param	context		    Any data provided by the user to serve as context on getDataCallback.
```
Yep

GeraldSchmitt commented on iothub_client/inc/blob.h:
```diff
@@ -39,15 +46,28 @@ DEFINE_ENUM(BLOB_RESULT, BLOB_RESULT_VALUES)
+* @param  httpStatus          A pointer to an out argument receiving the HTTP status (available only when the return value is BLOB_OK)
+* @param  httpResponse        A BUFFER_HANDLE that receives the HTTP response from the server (available only when the return value is BLOB_OK)
+*/
+MOCKABLE_FUNCTION(, BLOB_RESULT, Blob_UploadNextBlock, BUFFER_HANDLE, requestContent, unsigned int, blockID, STRING_HANDLE, xml, const char*, relativePath, HTTPAPIEX_HANDLE, httpApiExHandle, unsigned int*, httpStatus, BUFFER_HANDLE, httpResponse)
```
oops...

GeraldSchmitt commented on iothub_client/inc/iothub_client_ll.h:
```diff
@@ -549,6 +574,18 @@ extern "C"
+     *
+     * @return   IOTHUB_CLIENT_OK upon success or an error code upon failure.
+     */
+     MOCKABLE_FUNCTION(, IOTHUB_CLIENT_RESULT, IoTHubClient_LL_UploadMultipleBlocksToBlob, IOTHUB_CLIENT_LL_HANDLE, iotHubClientHandle, const char*, destinationFileName, IOTHUB_CLIENT_FILE_UPLOAD_GET_DATA_CALLBACK, getDataCallback, void*, context);
```
Done. I chose the shorter name.

GeraldSchmitt commented on iothub_client/src/iothub_client.c:
```diff
@@ -59,6 +59,15 @@ typedef struct UPLOADTOBLOB_SAVED_DATA_TAG
     int canBeGarbageCollected; /*flag indicating that the UPLOADTOBLOB_SAVED_DATA structure can be freed because the thread deadling with it finished*/
 }UPLOADTOBLOB_SAVED_DATA;
+
+typedef struct UPLOADMULTIPLEBLOCKS_DATA_TAG
```
Done

GeraldSchmitt commented on iothub_client/src/iothub_client.c:
```diff
@@ -1970,4 +1979,77 @@ IOTHUB_CLIENT_RESULT IoTHubClient_UploadToBlobAsync(IOTHUB_CLIENT_HANDLE iotHubC
+
+IOTHUB_CLIENT_RESULT IoTHubClient_UploadMultipleBlocksToBlobAsync(IOTHUB_CLIENT_HANDLE iotHubClientHandle, const char* destinationFileName, IOTHUB_CLIENT_FILE_UPLOAD_GET_DATA_CALLBACK iotHubClientFileUploadGetDataCallback, void* context)
+{
+    IOTHUB_CLIENT_RESULT result = IOTHUB_CLIENT_OK;
```
Done

GeraldSchmitt commented on iothub_client/src/iothub_client.c:
```diff
@@ -1970,4 +1979,77 @@ IOTHUB_CLIENT_RESULT IoTHubClient_UploadToBlobAsync(IOTHUB_CLIENT_HANDLE iotHubC
+        (iotHubClientFileUploadGetDataCallback == NULL)
+        )
+    {
+        LogError("invalid parameters IOTHUB_CLIENT_HANDLE iotHubClientHandle = %p , const char* destinationFileName = %s, IOTHUB_CLIENT_FILE_UPLOAD_GET_DATA_CALLBACK iotHubClientFileUploadGetDataCallback = %p, void* context = %p",
```
Done

GeraldSchmitt commented on iothub_client/src/iothub_client.c:
```diff
@@ -1970,4 +1979,77 @@ IOTHUB_CLIENT_RESULT IoTHubClient_UploadToBlobAsync(IOTHUB_CLIENT_HANDLE iotHubC
+            iotHubClientHandle,
+            destinationFileName,
+            iotHubClientFileUploadGetDataCallback,
+            context
```
Done

GeraldSchmitt commented on iothub_client/src/iothub_client.c:
```diff
@@ -1970,4 +1979,77 @@ IOTHUB_CLIENT_RESULT IoTHubClient_UploadToBlobAsync(IOTHUB_CLIENT_HANDLE iotHubC
+        UPLOADMULTIPLEBLOCKS_DATA *blocksData = (UPLOADMULTIPLEBLOCKS_DATA *)malloc(sizeof(UPLOADMULTIPLEBLOCKS_DATA));
+        if (blocksData == NULL)
+        {
+            LogError("unable to malloc - oom");
```
Out Of Memory, I got it from some other comments in the SDK.

ewertons replied:
I was almost sure it was there somewhere else in the repo. Sorry, we clean it as frequently as we can.

GeraldSchmitt commented on iothub_client/src/iothub_client.c:
```diff
@@ -1970,4 +1979,77 @@ IOTHUB_CLIENT_RESULT IoTHubClient_UploadToBlobAsync(IOTHUB_CLIENT_HANDLE iotHubC
+                blocksData->iotHubClientHandle = iotHubClientHandle;
+                if (ThreadAPI_Create(&blocksData->uploadingThreadHandle, uploadMultipleBlock_thread, blocksData) != THREADAPI_OK)
+                {
+                    /*Codes_SRS_IOTHUBCLIENT_02_053: [ If copying to the structure or spawning the thread fails, then IoTHubClient_UploadToBlobAsync shall fail and return IOTHUB_CLIENT_ERROR. ]*/
```
Yes, the requirement was a copy/paste from another block.
I did not pay much attention to writing the requirements yet, but I know it's still to be done. I'll ask some more questions on the email thread.

GeraldSchmitt commented on iothub_client/src/iothub_client_ll.c:
```diff
@@ -2073,4 +2073,26 @@ IOTHUB_CLIENT_RESULT IoTHubClient_LL_UploadToBlob(IOTHUB_CLIENT_LL_HANDLE iotHub
+        (getDataCallback == NULL)
+        )
+    {
+        LogError("invalid parameters IOTHUB_CLIENT_LL_HANDLE iotHubClientHandle=%p, const char* destinationFileName=%s, getDataCallback=%p, context=%p", iotHubClientHandle, destinationFileName, getDataCallback, context);
```
Done

GeraldSchmitt commented on iothub_client/src/iothub_client_ll_uploadtoblob.c:
```diff
@@ -71,6 +71,13 @@ typedef struct IOTHUB_CLIENT_LL_UPLOADTOBLOB_HANDLE_DATA_TAG
 
+typedef struct BLOB_UPLOAD_CONTEXT_TAG
+{
+    const unsigned char* source; /* source to upload */
```
I tried some more explicit names.

ewertons commented on iothub_client/devdoc/requirement_docs/iothubclient_ll_requirements.md:
```diff
@@ -503,56 +504,72 @@ IoTHubClient_LL_SetOption sets the runtime option "optionName" to the value poin
+### `IoTHubClient_LL_UploadToBlob` calls `IoTHubClient_LL_UploadToBlob_Impl` to synchronously upload the data pointed to by `source` having the size `size` to a blob called `destinationFileName` in Azure Blob Storage.
 
-**SRS_IOTHUBCLIENT_LL_02_061: [** If `iotHubClientHandle` is `NULL` then `IoTHubClient_LL_UploadToBlob` shall fail and return `IOTHUB_CLIENT_INVALID_ARG`.** ]**
+Design considerations: IoTHubClient_LL_UploadToBlob_Impl uses IoTHubClient_LL_UploadMultipleBlocksToBlob to upload the blob. An internal callback FileUpload_GetData_Callback and the corresponding context are used to chunk the source and feed it to IoTHubClient_LL_UploadMultipleBlocksToBlob.
```
Very nice. Thanks.

ewertons commented on iothub_client/tests/iothubclient_ll_ut/iothubclient_ll_ut.c:
```diff
@@ -3700,6 +3710,54 @@ TEST_FUNCTION(IoTHubClient_LL_UploadToBlob_with_NULL_source_and_size_greater_tha
     ///cleanup
     IoTHubClient_LL_Destroy(h);
 }
+
```
Are there requirement strings to be written for each of these tests? You know, the // Tests_SRS_...

GeraldSchmitt replied:
I should have seen this...

ewertons commented on iothub_client/tests/iothubclient_ut/iothubclient_ut.c:
```diff
@@ -2487,6 +2497,168 @@ TEST_FUNCTION(IoTHubClient_UploadToBlobAsync_succeeds)
     IoTHubClient_Destroy(iothub_handle);
 }
+
+/*SRS_IOTHUBCLIENT_99_072: [ If `iotHubClientHandle` is `NULL` then `IoTHubClient_UploadMultipleBlocksToBlobAsync` shall fail and return `IOTHUB_CLIENT_INVALID_ARG`. ]*/
```
Some of the requirement IDs are written without the "Tests_" prefix (to make them /* Tests_SRS_...)
Sorry, a little burden, I know, but that is part of an automation we have to check for requirement strings between code and tests.

GeraldSchmitt replied:
I should have seen this too! Btw, do you plan to release this kind automation tool?

GeraldSchmitt commented:
:champagne:
